arch/x86/entry.S

```diff
@@ -173,9 +181,15 @@ ENTRY(enter_usermode)
     PUSHF
 
     /* Save stack pointer onto per-cpu */
-    mov %_ASM_DX, %gs:(usermode_private)
+    mov %_ASM_SP, %gs:usermode_private
 
-    syscall_to_usermode
+    /* provide sysenter the kernel stack to be used */
+    push %_ASM_DX
+    WRITE_MSR $MSR_SYSENTER_ESP, %_ASM_SP
+    pop %_ASM_DX
```

Comment:
Do we really want to reprogram this MSR for each task? Why can't this simply be `percpu->tss.rsp0` and programmed just once in ?

Reply:
This is a good suggestion. Just make sure the `tss.rsp0` stack does not end up in the `usermode_private` var, as this is a placeholder for either kernel (main one) or user stack